### PR TITLE
Enable cl_khr_fp16 if the device can support it

### PR DESCRIPTION
**Pull request by jrprice**
This requires 16-bit storage and shaderFloat16.

### CODE REVIEW

File: src/device.cpp [hunk 1]
```diff
@@ -374,6 +374,14 @@ void cvk_device::build_extension_ils_list() {
             MAKE_NAME_VERSION(1, 0, 0, "cl_khr_device_uuid"));
     }
 
+    // Enable cl_khr_fp16 if we have 16-bit storage and shaderFloat16
+    if (is_vulkan_extension_enabled(VK_KHR_16BIT_STORAGE_EXTENSION_NAME) &&
+        is_vulkan_extension_enabled(
+            VK_KHR_SHADER_FLOAT16_INT8_EXTENSION_NAME) &&
```
Comment (kpet) [inline src/device.cpp:380]: These extensions have become core in 1.2. We need to handle that case as well.

File: src/device.cpp [hunk 2]
```diff
@@ -374,6 +374,14 @@ void cvk_device::build_extension_ils_list() {
...
+        m_features_float16_int8.shaderFloat16) {
```
Comment (kpet) [inline src/device.cpp:381]: I think we should check for at least `m_features_16bit_storage.storageBuffer16BitAccess` and probably `m_features_16bit_storage.uniformAndStorageBuffer16BitAccess` and/or  `m_features_16bit_storage.storagePushConstant16` for POD argument buffers and POD arguments in push constants.

**Review by kpet — CHANGES_REQUESTED** [review]

File: src/device.cpp (hunk 2 above)
Comment (jrprice) [reply src/device.cpp]: I've added a check for `storageBuffer16BitAccess`.

I don't think checking for uniforms and push constants is necessary. clvk already tells Clspv whether or not these things are supported, and Clspv should automatically fallback to storage buffers for things like POD arguments if necessary.

File: src/device.cpp (hunk 1 above)
Comment (jrprice) [reply src/device.cpp]: Can you clarify what needs to be done here? My impression (maybe incorrect?) was that these extensions would still be advertised even on newer Vulkan versions where they are core. I think the existing code for enabling extension during device initialization also makes this assumption; we only enable 16-bit storage features if the extension is present for example.

Comment (kpet) [reply src/device.cpp]: I could not find a definitive source but I'm pretty sure advertising the extensions is not a requirement. As you say, there are other cases so we might as well deal with all of them with one later change.

File: src/device.cpp (hunk 2 above)
Comment (kpet) [reply src/device.cpp]: Yes, with Alan's recent-ish changes, uniforms and push constants will be selected only when supported. I agree that checking for storage buffer support is enough.

**Review by kpet — APPROVED** [review]
LGTM, thanks!

**Pull request merged**